Token provider not called/ expired token not refreshed

### GetStream Environment
**GetStream Chat version: 2.2.9
**GetStream Chat frameworks:** StreamChatClient
**iOS version:** 13.x

Currently, we're using stream-chat-react-native for the front-end of our chat feature, this works great. We only use the StreamChatClient for swift to update unread counts on our native side so we only use it for subscription events. When the app is backgrounded and brought to the foreground after token expiry, the subscriptions break and the token provider is not called again. This is from the logs

```
[23 Aug 17:10:12.301] [DEBUG] ⬅️ Response 200 (0 bytes): https://chat-proxy-us-east.stream-io-api.com/connect?api_key=XXXXXXXXXX
[23 Aug 17:10:12.301] [INFO] 📦 ""
[23 Aug 17:10:12.301] [DEBUG] ♨️ TCP connection heated up.
[23 Aug 17:10:12.301] [ERROR] 📦  {
  "error" : {
    "code" : 40,
    "message" : "Connect failed with error: \"JWTAuth error: token is expired (exp)\"",
    "StatusCode" : 401,
    "duration" : "0.00ms"
  }
}
[23 Aug 17:10:12.302] [DEBUG] Parsing WebSocket disconnect... (error: The operation couldn’t be completed. (Starscream.WSError error 1.))
[23 Aug 17:10:12.302] [DEBUG] Clearing state after disconnect...
[23 Aug 17:10:12.302] [DEBUG] Cancelling background work...
[23 Aug 17:10:12.302] [DEBUG] Disconnected. 🀄️ Token is expired
```

After this the logs go silent and all subscriptions are silent. I assume that I need to make a request via the Client to trigger a token refresh, but we only use it for subscriptions. What would be the appropriate pattern to force a token refresh? Thanks!

gz#5502

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -11,6 +11,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 ### 🐞 Fixed
 - Channels list (ChannelsViewController) not updated when recreating a channel after deleting it [#392](https://github.com/GetStream/stream-chat-swift/issues/392)
+- If the user (JWT) token expires and websocket disconnects due to it, Client will renew the expiring JWT token and reconnect websocket automatically [#429](https://github.com/GetStream/stream-chat-swift/issues/429)
 
 # [2.2.9](https://github.com/GetStream/stream-chat-swift/releases/tag/2.2.9)
 _August 12, 2020_

--- a/Sources/Client/Client/Client+WebSocket.swift
+++ b/Sources/Client/Client/Client+WebSocket.swift
@@ -131,6 +131,10 @@ extension Client: WebSocketEventDelegate {
         // Don't clean up current user's typing events
         self.user != user
     }
+    
+    func disconnectedDueToExpiredToken() {
+        touchTokenProvider(isExpiredTokenInProgress: true, nil)
+    }
 }
 
 private struct WebSocketPayload: Encodable {

--- a/Sources/Client/WebSocket/WebSocket.swift
+++ b/Sources/Client/WebSocket/WebSocket.swift
@@ -236,6 +236,10 @@ protocol WebSocketEventDelegate: AnyObject {
     /// - Returns: A boolean value indicating whether a `typingStop` event should be sent for this user automatically
     ///   after the `incomingTypingStartEventTimeout` timeout.
     func shouldAutomaticallySendTypingStopEvent(for user: User) -> Bool
+    
+    /// Called after websocket disconnects due to invalid token.
+    /// The delegate should renew the token and call `connect()` on Websocket with a valid token.
+    func disconnectedDueToExpiredToken()
 }
 
 // MARK: - Web Socket Delegate
@@ -273,6 +277,7 @@ extension WebSocket: WebSocketProviderDelegate {
         if let eventError = eventError, eventError.code == ClientErrorResponse.tokenExpiredErrorCode {
             logger?.log("Disconnected. 🀄️ Token is expired")
             connectionStateAtomic.set(.disconnected(ClientError.expiredToken))
+            eventDelegate?.disconnectedDueToExpiredToken()
             return
         }
         

--- a/Tests/Client/Unit Tests/WebSocket/WebSocketTests.swift
+++ b/Tests/Client/Unit Tests/WebSocket/WebSocketTests.swift
@@ -162,4 +162,5 @@ private class MockDelegate: WebSocketEventDelegate {
     
     func shouldPublishEvent(_ event: Event) -> Bool { shouldPublishEvent }
     func shouldAutomaticallySendTypingStopEvent(for user: User) -> Bool { shouldAutomaticallySendTypingStopEvent }
+    func disconnectedDueToExpiredToken() { }
 }

